nav2_behavior_tree/plugins/decorator/distance_controller.cpp

```diff
@@ -0,0 +1,155 @@
+{
+public:
+  DistanceController(
+    const std::string & name,
+    const BT::NodeConfiguration & conf)
+  : BT::DecoratorNode(name, conf),
+    initialized_(false),
+    distance_(1.0)
+  {
+    getInput("distance", distance_);
+  }
+
+  // Any BT node that accepts parameters must provide a requiredNodeParameters method
+  static BT::PortsList providedPorts()
+  {
+    return {
+      BT::InputPort<double>("distance", 1.0, "Distance")
+    };
+  }
+
+  void initialize()
+  {
+    node_ = config().blackboard->get<rclcpp::Node::SharedPtr>("node");
+    tf_ = config().blackboard->get<std::shared_ptr<tf2_ros::Buffer>>("tf_buffer");
+    initialized_ = true;
+  }
+
+private:
+  BT::NodeStatus tick() override;
+
+  double euclidean_distance(
+    const geometry_msgs::msg::PoseStamped & pose1,
+    const geometry_msgs::msg::PoseStamped & pose2)
+  {
+    const double dx = pose1.pose.position.x - pose2.pose.position.x;
+    const double dy = pose1.pose.position.y - pose2.pose.position.y;
+    const double dz = pose1.pose.position.z - pose2.pose.position.z;
+    return std::sqrt(dx * dx + dy * dy + dz * dz);
+  }
+
+  rclcpp::Node::SharedPtr node_;
+
+  std::shared_ptr<tf2_ros::Buffer> tf_;
+
+  bool initialized_;
+
+  geometry_msgs::msg::PoseStamped start_pose_;
+  double distance_;
+};
+
+static bool first_time{false};
+
+inline BT::NodeStatus DistanceController::tick()
+{
+  if (!initialized_) {
+    initialize();
+  }
+
+  if (status() == BT::NodeStatus::IDLE) {
+    // Reset the starting position since we're starting a new iteration of
+    // the distance controller (moving from IDLE to RUNNING)
+    if (!nav2_util::getCurrentPose(start_pose_, *tf_)) {
+      RCLCPP_DEBUG(node_->get_logger(), "Current robot pose is not available.");
+      return BT::NodeStatus::FAILURE;
+    }
+    first_time = true;
+  }
+
+  setStatus(BT::NodeStatus::RUNNING);
+
+  // Determine distance travelled since we've started this iteration
+  geometry_msgs::msg::PoseStamped current_pose;
+  if (!nav2_util::getCurrentPose(current_pose, *tf_)) {
+    RCLCPP_DEBUG(node_->get_logger(), "Current robot pose is not available.");
+    return BT::NodeStatus::FAILURE;
+  }
+
+  // Get euclidean distance
+  auto travelled = euclidean_distance(start_pose_, current_pose);
+
+  // The child gets ticked the first time through and every time the threshold
+  // distance is crossed. In addition, once the child begins to run, it is
+  // ticked each time 'til completion
+  if (first_time || (child_node_->status() == BT::NodeStatus::RUNNING) ||
+    travelled >= distance_)
+  {
+    first_time = false;
+    const BT::NodeStatus child_state = child_node_->executeTick();
+
+    switch (child_state) {
+      case BT::NodeStatus::RUNNING:
+        return BT::NodeStatus::RUNNING;
+
+      case BT::NodeStatus::SUCCESS:
+        child_node_->setStatus(BT::NodeStatus::IDLE);
+        // reset starting pose to current pose
+        if (!nav2_util::getCurrentPose(start_pose_, *tf_)) {
```

Comment:
Why not just set first_time to true again to cause the same update?

Reply:
Setting ```first_time``` to true there would make the controller satisfy the replanning condition at every tick. ```first_time``` is just a flag to ensure that planning happens at the very first tick when motion has not actually started. The only time it should be set to true is after the node is transitioning from IDLE to RUNNING state.

Instead of setting ```first_time``` to true, the current pose is recorded at this point when the child node ticks successfully so that the condition can be checked in the next ticks.

Reply:
OK, I think it could be refactored but this is totally fine.